sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestStream.java

```diff
@@ -0,0 +1,326 @@
+      events.add(WatermarkEvent.<T>advanceTo(newWatermark));
+      currentWatermark = newWatermark;
+      return this;
+    }
+
+    /**
+     * Advance the processing time by the specified amount.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceProcessingTime(Duration amount) {
+      checkArgument(
+          amount.getMillis() > 0,
+          "Must advance the processing time by a positive amount. Got: ",
+          amount);
+      events.add(ProcessingTimeEvent.<T>advanceBy(amount));
+      return this;
+    }
+
+    /**
+     * Advance the watermark to infinity, completing this {@link TestStream}. Future calls to the
+     * same builder will not affect the returned {@link TestStream}.
+     */
+    public TestStream<T> advanceWatermarkToInfinity() {
+      events.add(WatermarkEvent.<T>advanceTo(BoundedWindow.TIMESTAMP_MAX_VALUE));
+      return new TestStream<>(coder, events.build());
+    }
+  }
+
+  /**
+   * An event in a {@link TestStream}. A marker interface for all events that happen while
+   * evaluating a {@link TestStream}.
+   */
+  public interface Event<T> {
+    EventType getType();
+  }
+
+  /**
+   * The types of {@link Event} that are supported by {@link TestStream}.
+   */
+  public enum EventType {
+    ELEMENT,
+    WATERMARK,
+    PROCESSING_TIME
+  }
+
+  /** A {@link Event} that produces elements. */
+  @AutoValue
+  public abstract static class ElementEvent<T> implements Event<T> {
+    public abstract Iterable<TimestampedValue<T>> getElements();
+
+    @SafeVarargs
+    static <T> Event<T> add(TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      return add(ImmutableList.<TimestampedValue<T>>builder().add(element).add(elements).build());
+    }
+
+    static <T> Event<T> add(Iterable<TimestampedValue<T>> elements) {
+      return new AutoValue_TestStream_ElementEvent<>(EventType.ELEMENT, elements);
+    }
+  }
+
+  /** A {@link Event} that advances the watermark. */
+  @AutoValue
+  public abstract static class WatermarkEvent<T> implements Event<T> {
+    public abstract Instant getWatermark();
+
+    static <T> Event<T> advanceTo(Instant newWatermark) {
+      return new AutoValue_TestStream_WatermarkEvent<>(EventType.WATERMARK, newWatermark);
+    }
+  }
+
+  /** A {@link Event} that advances the processing time clock. */
+  @AutoValue
+  public abstract static class ProcessingTimeEvent<T> implements Event<T> {
+    public abstract Duration getProcessingTimeAdvance();
+
+    static <T> Event<T> advanceBy(Duration amount) {
+      return new AutoValue_TestStream_ProcessingTimeEvent<>(EventType.PROCESSING_TIME, amount);
+    }
+  }
+
+  @Override
+  public PCollection<T> apply(PBegin input) {
+    return PCollection.<T>createPrimitiveOutputInternal(
+            input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED)
+        .setCoder(coder);
+  }
+
+  public List<Event<T>> getStreamEvents() {
```

Comment:
Add a comment that this method is meant for runner authors as an implementation detail of the test stream
getStreamEvents() -> getEvents()